File: app/src/main/java/com/example/locationmocker/domain/track/TrackDetection.kt
package com.example.locationmocker.domain.track

import com.example.locationmocker.domain.model.RoutePoint
import com.example.locationmocker.domain.route.RouteMath

data class TrackCandidate(
    val name: String,
    val center: RoutePoint,
    val typeDescription: String = "",
    val distanceMeters: Double = 0.0,
)

sealed interface TrackDetectionResult {
    data class Success(
        val name: String,
        val center: RoutePoint,
        val confidence: Int,
        val sourceBounds: List<RoutePoint> = emptyList(),
    ) : TrackDetectionResult

    data class NotFound(val reason: String) : TrackDetectionResult
}

object TrackCandidateScorer {
    private val strongKeywords = listOf(
        "\u64cd\u573a",
        "\u7530\u5f84\u573a",
        "\u8fd0\u52a8\u573a",
        "\u8dd1\u9053",
    )
    private val weakKeywords = listOf(
        "\u4f53\u80b2\u573a",
        "\u4f53\u80b2\u4e2d\u5fc3",
        "\u8fd0\u52a8\u4e2d\u5fc3",
        "\u8db3\u7403\u573a",
    )

    fun bestCandidate(origin: RoutePoint, candidates: List<TrackCandidate>): TrackDetectionResult {
        val scored = candidates.mapNotNull { candidate ->
            val score = score(origin, candidate)
            if (score >= 55) candidate to score else null
        }.maxByOrNull { it.second }

        return if (scored == null) {
            TrackDetectionResult.NotFound("\u9644\u8fd1\u672a\u8bc6\u522b\u5230\u64cd\u573a")
        } else {
            val candidate = scored.first
            TrackDetectionResult.Success(
                name = candidate.name.ifBlank { "\u9644\u8fd1\u64cd\u573a" },
                center = origin,
                confidence = scored.second,
            )
        }
    }

    fun score(origin: RoutePoint, candidate: TrackCandidate): Int {
        val text = "${candidate.name} ${candidate.typeDescription}"
        val keywordScore = when {
            strongKeywords.any { text.contains(it) } -> 70
            weakKeywords.any { text.contains(it) } -> 45
            else -> 0
        }
        if (keywordScore == 0) return 0

        val distance = candidate.distanceMeters.takeIf { it > 0.0 }
            ?: RouteMath.distanceMeters(origin, candidate.center)
        val distanceScore = when {
            distance <= 80.0 -> 25
            distance <= 150.0 -> 18
            distance <= 250.0 -> 10
            else -> -20
        }
        return (keywordScore + distanceScore).coerceIn(0, 100)
    }

}

File: ios/LocationMocker/LocationMocker/Domain/Track/TrackDetection.swift
import Foundation

struct TrackCandidate {
    let name: String
    let center: RoutePoint
    let typeDescription: String
    let distanceMeters: Double

    init(name: String, center: RoutePoint, typeDescription: String = "", distanceMeters: Double = 0) {
        self.name = name
        self.center = center
        self.typeDescription = typeDescription
        self.distanceMeters = distanceMeters
    }
}

enum TrackDetectionResult {
    case success(name: String, center: RoutePoint, confidence: Int, sourceBounds: [RoutePoint])
    case notFound(reason: String)
}

enum TrackCandidateScorer {
    private static let strongKeywords = ["操场", "田径场", "运动场", "跑道"]
    private static let weakKeywords = ["体育场", "体育中心", "运动中心", "足球场"]

    static func bestCandidate(origin: RoutePoint, candidates: [TrackCandidate]) -> TrackDetectionResult {
        let scored = candidates.compactMap { candidate -> (TrackCandidate, Int)? in
            let s = score(origin: origin, candidate: candidate)
            return s >= 55 ? (candidate, s) : nil
        }
        guard let best = scored.max(by: { $0.1 < $1.1 }) else {
            return .notFound(reason: "附近未识别到操场")
        }
        let name = best.0.name.isEmpty ? "附近操场" : best.0.name
        return .success(name: name, center: origin, confidence: best.1, sourceBounds: [])
    }

    static func score(origin: RoutePoint, candidate: TrackCandidate) -> Int {
        let text = "\(candidate.name) \(candidate.typeDescription)"
        let keywordScore: Int
        if strongKeywords.contains(where: text.contains) {
            keywordScore = 70
        } else if weakKeywords.contains(where: text.contains) {
            keywordScore = 45
        } else {
            return 0
        }
        let distance = candidate.distanceMeters > 0
            ? candidate.distanceMeters
            : RouteMath.distanceMeters(origin, candidate.center)
        let distanceScore: Int
        if distance <= 80 { distanceScore = 25 }
        else if distance <= 150 { distanceScore = 18 }
        else if distance <= 250 { distanceScore = 10 }
        else { distanceScore = -20 }
        return min(max(keywordScore + distanceScore, 0), 100)
    }
}
